project-rainstorm/project_rainstorm
# Implement JWT Refresh Token

# What was done 
- Updated our JWT library to be flask_jwt_extended since it came with refresh tokens out of the box
- Updated the login route to return an access_token and a refresh_token 
- Updated login logic to read from the app config for username and password
- Added a refresh api call method to the AuthService class
- Updated logic for logout button to delete tokens from storage instead of just sending you to the login screen
- Updated Login component to automatically try and use a refresh token to re-authenticate if one is present 

# How To Test
1. Checkout the branch on your pi 
2. Within your python virutal env, run pip3 install -r requriements.txt again to download flask_jwt_extended
2. Edit line 24 of __init__.py for the flask server to be 5 seconds instead of 900 (so you don't have to wait 15 min to verify the refresh logic)
```
JWT_ACCESS_TOKEN_EXPIRES=timedelta(seconds=900),  # 15 min
```
2. Start FED and BED dev servers 
3. Logout on the frontend (in case you already have an access_token stored)
4.  Login using the drop user name and password 
5. Wait 5 seconds 
6. Hit refresh 
7. You should still be logged in
8. Verify the server logs show you were unauthenticated but it automatically refreshed your access_token

![image](https://user-images.githubusercontent.com/16943263/99075389-6299ee80-2587-11eb-98e3-4c2a4302cb4a.png)
